Sources/SwiftDiscord/DiscordEndpoint.swift

```diff
@@ -27,149 +44,381 @@ import Foundation
     /// The user channels endpoint.
-    case userChannels = "/users/me/channels"
+    case userChannels
 
     /// The user guilds endpoint.
-    case userGuilds = "/users/me/guilds"
+    case userGuilds
 
     /* Webhooks */
     /// The single webhook endpoint.
-    case webhook = "/webhooks/webhook.id"
+    case webhook(id: String)
 
     /// The single webhook with token endpoint.
-    case webhookWithToken = "/webhooks/webhook.id/webhook.token"
+    case webhookWithToken(id: String, token: String)
 
     /// A slack compatible webhook.
-    case webhookSlack = "/webhooks/webhook.id/webhook.token/slack"
+    case webhookSlack(id: String, token: String)
 
     /// A GitHub compatible webhook.
-    case webhookGithub = "/webhooks/webhook.id/webhook.token/github"
+    case webhookGithub(id: String, token: String)
     /* End Webhooks */
 
     var combined: String {
-        return DiscordEndpoint.baseURL.rawValue + rawValue
+        return DiscordEndpoint.baseURL.description + description
     }
 
-    // MARK: Methods
-
-    /**
-        Helper method that creates the basic request for an endpoint.
-
-        - parameter with: A DiscordToken that will be used for authentication
-        - parameter for: The endpoint this request is for
-        - parameter replacing: A dictionary that will be used to fill in the endpoint's url
-        - parameter getParams: An optional dictionary of get parameters.
-
-        - returns: a URLRequest that can be further customized
-    */
-    public static func createRequest(with token: DiscordToken, for endpoint: DiscordEndpoint,
-                                     replacing: [String: String], getParams: [String: String]? = nil) -> URLRequest {
-
-        var request = URLRequest(url: endpoint.createURL(replacing: replacing, getParams: getParams ?? [:]))
+	// MARK: Endpoint Request enum
+
+	/**
+	* An HTTP Request for an Endpoint.  This includes any associated data.
+	*/
+	public enum EndpointRequest {
+		case get(params: [String: String]?)
+		case post(content: (Data, type: HTTPContentType)?)
+		case put(content: (Data, type: HTTPContentType)?)
+		case patch(content: (Data, type: HTTPContentType)?)
+		case delete
+
+		var methodString: String {
+			switch self {
+			case .get:
+				return "GET"
+			case .post:
+				return "POST"
+			case .put:
+				return "PUT"
+			case .patch:
+				return "PATCH"
+			case .delete:
+				return "DELETE"
+			}
+		}
+
+		/**
+		Helper method that creates the basic request for an endpoint.
+
+		- parameter with: A DiscordToken that will be used for authentication
+		- parameter for: The endpoint this request is for
+		- parameter getParams: An optional dictionary of get parameters.
+
+		- returns: a URLRequest that can be further customized
+		*/
+		public func createRequest(with token: DiscordToken, endpoint: DiscordEndpoint) -> URLRequest? {
+
+			let getParams: [String: String]?
+			if case let .get(params) = self {
+				getParams = params
+			}
+			else {
```

Comment:
Sorry, but I'm a bit nitpicky when it comes style, esp with things like brackets.